File: bench/langs/sortnums.py
"""Sort N pseudo-random integers, Reps times -- the same recurrence."""
import sys

def main(n, reps):
    c = 0
    for _ in range(reps):
        l = []
        x = 12345
        for _i in range(n):
            x = (1103515245 * x + 12345) % 2147483648
            l.append(x % 100000)
        a = 0
        for h in sorted(l):
            a = (a * 31 + h) % 1000003
        c = a
    print("answer(%d)" % c)

if __name__ == "__main__":
    main(int(sys.argv[1]), int(sys.argv[2]))

File: bench/langs/sortnums.pl
%% Sort N pseudo-random integers. Both sides generate the SAME numbers
%% from the same linear congruential recurrence, so the answer checksums
%% agree only if both sorted the same list.
%%
%%   main(+N)   answer is a checksum over the sorted list

lcg(0, _, []) :- !.
lcg(K, X0, [X|T]) :-
    X1 is (1103515245 * X0 + 12345) mod 2147483648,
    X is X1 mod 100000,
    K1 is K - 1,
    lcg(K1, X1, T).

sum31([], A, A).
sum31([H|T], A0, A) :- A1 is (A0 * 31 + H) mod 1000003, sum31(T, A1, A).

%% One rep is one generate-and-sort, and every rep answers the same
%% checksum.
reps(0, _, C, C) :- !.
reps(K, N, _, C) :-
    lcg(N, 12345, L),
    msort(L, S),
    sum31(S, 0, C0),
    K1 is K - 1,
    reps(K1, N, C0, C).

main(N, Reps) :-
    reps(Reps, N, 0, C),
    format("answer(~w)~n", [C]).
